### PR TITLE
Sg/swizzle node UI alignment

### PR DESCRIPTION
**Pull request by tracychen1234**
# **Please read the [Contributing guide](CONTRIBUTING.md) before making a PR.**

* Read the [Graphics repository & Yamato FAQ](http://go/graphics-yamato-faq).

### Checklist for PR maker
- [x] Have you added a backport label (if needed)? For example, the `need-backport-*` label. After you backport the PR, the label changes to `backported-*`.
- [x] Have you updated the changelog? Each package has a `CHANGELOG.md` file.
- [x] Have you updated or added the documentation for your PR? When you add a new feature, change a property name, or change the behavior of a feature, it's best practice to include related documentation changes in the same PR. If you do add documentation, make sure to add the relevant Graphics Docs team member as a reviewer of the PR. If you are not sure which person to add, see the [Docs team contacts sheet](https://docs.google.com/spreadsheets/d/1rgUWWgwLFEHIQ3Rz-LnK6PAKmbM49DZZ9al4hvnztOo/edit#gid=1058860420).
- [x] Have you added a graphic test for your PR (if needed)? When you add a new feature, or discover a bug that tests don't cover, please add a graphic test.

---
### Purpose of this PR
Why is this PR needed, what hard problem is it solving/fixing?
Update current Swizzle node UI for better alignment with other graph tools.
image

---
### Testing status
Describe what manual/automated tests were performed for this PR
Tested with different vector size input values and mask input edge cases.
Succeed in both local Paly Mode/Edit Mode tests (reconnected the swizzle node in tests).

---
### Comments to reviewers
Notes for the reviewers you have assigned.

### CODE REVIEW

**Review by LandonTownsendUnity — CHANGES_REQUESTED** [review]
We can't just replace nodes like this - a shadergraph using the old version of the swizzle node will not work correctly when upgraded to this branch. Either the node needs to upgrade to an equivalent version of the new node (which may be difficult to code and to test) or another solution may be making it an entirely new node, like "Swizzle inline". Since this is feature work, we may want to discuss how we should do this.

File: com.unity.shadergraph/Editor/Data/Nodes/Channel/SwizzleNode.cs [hunk 1]
```diff
@@ -22,151 +24,105 @@ public SwizzleNode()
...
+        //2.mask cant be character other than "xyzw"
+        //3.If the input is not valid, wont genetate shader code and give errors
+        public bool ValidateMaskInput(int InputValueSize)
         {
-            get { return m_GreenChannel; }
-            set
+            bool MaskInputIsValid = true;
+            char[] MaskChars = _maskInput.ToCharArray();
+            char[] AllChars  = {'x', 'y' , 'z', 'w'};
+            List<char> CurrentChars = new List<char>();
```
Comment (cdxntchou) [inline com.unity.shadergraph/Editor/Data/Nodes/Channel/SwizzleNode.cs:53]: Instead of copying from the AllChars arrayr into the CurrentChars list, you can simply initialize the CurrentChars list to be what you want:
List<char> CurrentChars = new List<char>() { 'x', 'y', 'z', 'w' };

**Review by LandonTownsendUnity — CHANGES_REQUESTED** [review]
Swizzle node also lacks a preview now, which the earlier version had:
image

File: com.unity.shadergraph/Editor/Data/Nodes/Channel/SwizzleNode.cs [hunk 2]
```diff
@@ -22,151 +24,105 @@ public SwizzleNode()
...
-
-        [SerializeField]
-        TextureChannel m_AlphaChannel;
-
-        [ChannelEnumControl("Alpha Out")]
-        public TextureChannel alphaChannel
-        {
-            get { return m_AlphaChannel; }
-            set
+            if (MaskChars.Length == 0 || MaskChars.Length > 4)
```
Comment (cdxntchou) [inline com.unity.shadergraph/Editor/Data/Nodes/Channel/SwizzleNode.cs:67]: Even though it's not strictly necessary for Length, I generally like to do tests that will catch errors for any possible value.  So in this case:
if (MaskChars.Length <= 0 || MaskChars.Length > 4)

**Comment by tracychen1234** [comment]
> We can't just replace nodes like this - a shadergraph using the old version of the swizzle node will not work correctly when upgraded to this branch. Either the node needs to upgrade to an equivalent version of the new node (which may be difficult to code and to test) or another solution may be making it an entirely new node, like "Swizzle inline". Since this is feature work, we may want to discuss how we should do this.

Working on backward compatibilities

**Comment by tracychen1234** [comment]
> Swizzle node also lacks a preview now, which the earlier version had:
> image

Took out the preview window for better alignment with vfx graph, also we now have a Preview node so I think it's fine to take it off?

**Comment by LandonTownsendUnity** [comment]
> > Swizzle node also lacks a preview now, which the earlier version had:
> > image
> 
> Took out the preview window for better alignment with vfx graph, also we now have a Preview node so I think it's fine to take it off?

I think VFX graph lacks a preview because the nature of how VFX graph works, node previews don't really work... but I think the node should have a preview in Shadergraph

File: com.unity.shadergraph/Editor/Data/Nodes/Channel/SwizzleNode.cs [hunk 3]
```diff
@@ -22,151 +24,107 @@ public SwizzleNode()
...
-            });
-        }
-
-        public override void CollectPreviewMaterialProperties(List<PreviewProperty> properties)
-        {
-            base.CollectPreviewMaterialProperties(properties);
-            // Encode swizzle values into an integer
-            var value = ((int)redChannel) | ((int)greenChannel << 2) | ((int)blueChannel << 4) | ((int)alphaChannel << 6);
-            properties.Add(new PreviewProperty(PropertyType.Float)
+                owner.AddValidationError(objectId, "Invalid mask!", ShaderCompilerMessageSeverity.Error);
```
Comment (marctem) [inline com.unity.shadergraph/Editor/Data/Nodes/Channel/SwizzleNode.cs:109]: No exclamation points in error messages please. It's not exciting =). Can we give them a hint what would be valid instead? (Like, "Invalid Mask. Valid values are r, g, b.")

File: com.unity.shadergraph/Documentation~/Swizzle-Node.md [hunk 4]
```diff
@@ -17,15 +18,12 @@ Channel dropdown parameters are dynamic depending on the length of the input vec
...
 
 ## Generated Code Example
 
 The following example code represents one possible outcome of this node.
 
 ```
-float4 _Swizzle_Out = In.xzyw;
-```
\ No newline at end of file
+float4 _Swizzle_Out = float4 (In.y,In.z,In.w,In.x);
```
Comment (marctem) [inline com.unity.shadergraph/Documentation~/Swizzle-Node.md:29]: Is the float cast here slower than the old explicit swizzle code?

File: com.unity.shadergraph/Editor/Data/Nodes/Channel/SwizzleNode.cs [hunk 5]
```diff
@@ -22,151 +24,107 @@ public SwizzleNode()
...
+                sb.AppendLine(string.Format("{0} {1} = float4 (0, 0, 0, 0);", outputSlotType, outputName));
```
Comment (marctem) [inline com.unity.shadergraph/Editor/Data/Nodes/Channel/SwizzleNode.cs:110]: Does this work with the precision controls?

File: com.unity.shadergraph/Editor/Data/Nodes/Channel/SwizzleNode.cs [hunk 6]
```diff
@@ -22,151 +24,107 @@ public SwizzleNode()
...
+                    if (i != 0)
+                        outputValue += ",";
+
+                    if (i < _maskInput.Length)
+                        outputValue += inputValue + "." + _maskInput[i];
+
+                    if (i >= _maskInput.Length)
+                        outputValue += "0";
+                }
+                sb.AppendLine("{0} {1} = float4 ({2});", outputSlotType, outputName, outputValue);
```
Comment (marctem) [inline com.unity.shadergraph/Editor/Data/Nodes/Channel/SwizzleNode.cs:126]: Same question about precision.

**Review by marctem — CHANGES_REQUESTED** [review]

File: com.unity.shadergraph/Documentation~/Swizzle-Node.md [hunk 7]
```diff
@@ -17,15 +18,12 @@ Channel dropdown parameters are dynamic depending on the length of the input vec
 
 | Name        | Type           | Options  | Description |
 |:------------ |:-------------|:-----|:---|
-| Red out      | Dropdown | Red, Green, Blue, Alpha (depending on input vector dimension) | Defines which input channel should be used in the output's red channel |
-| Green out      | Dropdown | Red, Green, Blue, Alpha (depending on input vector dimension) | Defines which input channel should be used in the output's green channel |
-| Blue out      | Dropdown | Red, Green, Blue, Alpha (depending on input vector dimension) | Defines which input channel should be used in the output's blue channel |
-| Alpha out      | Dropdown | Red, Green, Blue, Alpha (depending on input vector dimension) | Defines which input channel should be used in the output's alpha channel |
+| Mask     | Inputfield | X, Y, Z, W (depending on input vector dimension) | The swizzle mask is a combination of one to four characters that can be x, y, z, or w. The size of output value depends on the length of the mask input.|
```
Comment (Nightmask3) [inline com.unity.shadergraph/Documentation~/Swizzle-Node.md:21]: I'm also able to use r, g, b, a as mask inputs. So would add a mention of that here.

**Review by Nightmask3 — APPROVED** [review]
This looks good! Tested it out a decent amount and can't find any issues.

Also really like how the new swizzle node works, can totally see myself using it a lot. 
Really good stuff Tracy! 👍

**Comment by Nightmask3** [comment]
> > > Swizzle node also lacks a preview now, which the earlier version had:
> > > image
> > 
> > 
> > Took out the preview window for better alignment with vfx graph, also we now have a Preview node so I think it's fine to take it off?
> 
> I think VFX graph lacks a preview because the nature of how VFX graph works, node previews don't really work... but I think the node should have a preview in Shadergraph

I do agree that an on-the-same-node preview would be really good to have for quick preview of what the swizzle operation output was, as otherwise I'd have to connect to a graph output or preview node which adds a few more clicks. 

While testing this out I found myself wanting a preview to be able to confirm that the values were in fact changing as I modified the mask.

Buuut we do also have a preview node for a reason I guess so I can see arguments for going either way. 

I guess that does raise the question as to why we would even have a preview node if all nodes had their own previews, so thats a potential reason to exclude the preview from the swizzle node. 

The previews do add some visual noise to the graph as a whole, so that's another motivation to exclude them.

**Comment by tracychen1234** [comment]
> > > > Swizzle node also lacks a preview now, which the earlier version had:
> > > > image
> > > 
> > > 
> > > Took out the preview window for better alignment with vfx graph, also we now have a Preview node so I think it's fine to take it off?
> > 
> > 
> > I think VFX graph lacks a preview because the nature of how VFX graph works, node previews don't really work... but I think the node should have a preview in Shadergraph
> 
> I do agree that an on-the-same-node preview would be really good to have for quick preview of what the swizzle operation output was, as otherwise I'd have to connect to a graph output or preview node which adds a few more clicks.
> 
> While testing this out I found myself wanting a preview to be able to confirm that the values were in fact changing as I modified the mask.
> 
> Buuut we do also have a preview node for a reason I guess so I can see arguments for going either way.
> 
> I guess that does raise the question as to why we would even have a preview node if all nodes had their own previews, so thats a potential reason to exclude the preview from the swizzle node.
> 
> The previews do add some visual noise to the graph as a whole, so that's another motivation to exclude them.
> 
> Ideally we'd just be able to toggle previews on every node as the user needed so maybe we could just post-pone this discussion till we can circle back on a more graph-wide convention.

Added preview. Thanks for the feedback, Sai! I didn't realize that VFX graph lacking preview is because of their nature, thanks Landon! The preview on the node does serve as a role of providing positive/negative feedback and I feel it's quite important for a node like swizzle to give visual feedback. And I also agree with Sai that too many previews might cause some visual noise. But the user can always collapse the on-node preview window (sounds less effort comparing to plug in a preview node every time). Also asked Alex, she's good with adding a preview.

File: com.unity.shadergraph/Editor/Data/Nodes/Channel/SwizzleNode.cs (hunk 5 above)
Comment (tracychen1234) [reply com.unity.shadergraph/Editor/Data/Nodes/Channel/SwizzleNode.cs:109]: Good point! Changed to `sb.AppendLine(string.Format("{0} {1} = 0;", outputSlotType, outputName));`

File: com.unity.shadergraph/Editor/Data/Nodes/Channel/SwizzleNode.cs (hunk 3 above)
Comment (tracychen1234) [reply com.unity.shadergraph/Editor/Data/Nodes/Channel/SwizzleNode.cs]: Took off the exclamation mark. And for the error message what do you think about this?
Screenshot (4)

**Review by LandonTownsendUnity — CHANGES_REQUESTED** [review]
Needs to validate that you are not mixing and matching rgba with xyzw. After recent changes, doing this causes compile errors. Also, I think it might be a nice improvement to usability to automatically convert the input string to lowercase so that users won't be confused if they try something like "XYZ" and it doesn't work.
image

**Comment by tracychen1234** [comment]
https://github.com/Unity-Technologies/Graphics/pull/3040 is the correct one
